src/validation.cpp

```diff
@@ -3430,48 +3430,87 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
+void CChainState::ProcessBlockValidationQueue()
 {
-    AssertLockNotHeld(cs_main);
+    while (true) {
+        LimitValidationInterfaceQueue();
 
-    std::promise<bool> result_promise;
-    std::future<bool> result = result_promise.get_future();
-    bool fNewBlock = false;
+        std::shared_ptr<const CBlock> pblock;
+        bool fForceProcessing;
+        std::promise<bool> result_promise;
+        {
+            std::unique_lock<CCriticalSection> lock(m_cs_block_validation_queue);
+            if (m_block_validation_queue.empty()) {
+                m_cv_block_validation_queue.wait_for(lock, std::chrono::seconds(1));
+            }
+            if (ShutdownRequested())
+                break;
+            boost::this_thread::interruption_point();
+            if (m_block_validation_queue.empty()) {
+                continue;
+            }
 
-    {
-        CBlockIndex *pindex = nullptr;
-        CValidationState state;
+            std::tuple<std::shared_ptr<const CBlock>, bool, std::promise<bool>>& tuple = m_block_validation_queue.front();
+            pblock = std::move(std::get<0>(tuple));
+            fForceProcessing = std::get<1>(tuple);
+            result_promise = std::move(std::get<2>(tuple));
+            m_block_validation_queue.pop_front();
+        }
 
-        // CheckBlock() does not support multi-threaded block validation because CBlock::fChecked can cause data race.
-        // Therefore, the following critical section must include the CheckBlock() call as well.
-        LOCK(cs_main);
+        CChainParams chainparams = Params();
+        bool fNewBlock = false;
+        {
+            CBlockIndex *pindex = nullptr;
+            CValidationState state;
 
-        // Ensure that CheckBlock() passes before calling AcceptBlock, as
-        // belt-and-suspenders.
-        bool ret = CheckBlock(*pblock, state, chainparams.GetConsensus());
-        if (ret) {
-            // Store to disk
-            ret = ::ChainstateActive().AcceptBlock(pblock, state, chainparams, &pindex, fForceProcessing, nullptr, &fNewBlock);
-        }
-        if (!ret) {
-            GetMainSignals().BlockChecked(*pblock, state);
-            error("%s: AcceptBlock FAILED (%s)", __func__, FormatStateMessage(state));
-            result_promise.set_value(fNewBlock);
-            return result;
+            // CheckBlock() does not support multi-threaded block validation because CBlock::fChecked can cause data race.
+            // Therefore, the following critical section must include the CheckBlock() call as well.
+            LOCK(cs_main);
+
+            // Ensure that CheckBlock() passes before calling AcceptBlock, as
+            // belt-and-suspenders.
+            bool ret = CheckBlock(*pblock, state, chainparams.GetConsensus());
+            if (ret) {
+                // Store to disk
+                ret = ::ChainstateActive().AcceptBlock(pblock, state, chainparams, &pindex, fForceProcessing, nullptr, &fNewBlock);
```

Comment:
In commit "Run ProcessNewBlock in a background thread (XXX: p2p_sendheaders broken)" (64e74ce8a917e071b8a250a4ba9cd110448796d6)

Throughout this function shouldn't `::ChainstateActive()` be replaced by `*this`?